plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java

```diff
@@ -699,7 +700,9 @@ public ConnectorInsertTableHandle beginInsert(ConnectorSession session, Connecto
                 .map(JdbcColumnHandle.class::cast)
                 .collect(toImmutableList());
         JdbcOutputTableHandle handle = jdbcClient.beginInsertTable(session, (JdbcTableHandle) tableHandle, columnHandles);
-        setRollback(() -> jdbcClient.rollbackCreateTable(session, handle));
+        if (!isNonTransactionalInsert(session)) {
+            setRollback(() -> jdbcClient.rollbackCreateTable(session, handle));
+        }
```

Comment:
This should be a logic of `jdbcClient.rollbackCreateTable`
since `isNonTransactionalInsert` is handled by `jdbcClient.beginInsertTable`

(eg imagine a connector which overrides `jdbcClient.beginInsertTable` and ignores `isNonTransactionalInsert` session property; it would have no way to behave correctly)

Reply:
What if someone calls jdbcClient.rollbackCreateTable in some other context? In that case we'd want to remove that table regardless of whether inserts are transactional or not.

Reply:
Seem it's wrong either way:
DefaultJdbcMetadata.beginCreateTable also sets up jdbcClient.rollbackCreateTable as rollback action. In that case we always want the rollback to happen.
For DefaultJdbcMetadata.beginInsert we want the rollback handled conditionally.

Reply:
i think `JdbcOutputTableHandle` needs to track information whether this is a temp table or not.

Reply:
Actually... Even that would not work (nicely). It would need to be a specific flag for isNonTransactionalInsert.
All other table creation (that we want to have removed) have to be marked as temporary. Now you have remember explicitly not do that, seems more frail to me. IMHO this has to be connected to an action, not the OutputTable.

I think my initial fix is still the safest. BaseJdbcClient.beginInsertTable is not overriden (except for disabling via throwing an Exception). And when one overrides that, one would have to aware of the implications. One could put a comment on that method.

As is, this will happily drop your TB size table if you fail to insert into it on non-transactional mode. :(

IMHO, the damage is done in the original implementation that separated table creation from the responsibility to clean the table up into two different classes. It's not immediately clear how to undo that.

Reply:
> I think my initial fix is still the safest.

it spreads responsibility between the two classes: DefaultJdbcMetadata (which generally should not be overridden)  and `JdbcClient` which is the extension point for all the jdbc-based connectors

> All other table creation (that we want to have removed) have to be marked as temporary. Now you have remember explicitly not do that, seems more frail to me

I thought about adding new parameter to `JdbcOutputTableHandle`. You cannot forget about it, because it's a required parameter.

However, we don't need a new parameter, we should just compare `JdbcOutputTableHandle#tableName` and `JdbcOutputTableHandle#temporaryTableName`. Or, better, make `temporaryTableName` Optional.

Reply:
> it spreads responsibility between the two classes

That already happened because BaseJdbcClient creates the tables, but DefaultJdbcMetadata is responsible for the cleanup.

> I thought about adding new parameter to JdbcOutputTableHandle. You cannot forget about it, because it's a required parameter.

I started with that, easy enough to do. But I ran into the issue I mentioned above.

> compare JdbcOutputTableHandle#tableName and JdbcOutputTableHandle#temporaryTableName

So from looking `BaseJdbcClient.beginInsertTable` you are suggesting that `DefaultJdbcMetadata.beginInsert` to set up the rollback only handler when tableName and temporaryTableName are identical? (real tables seem to have these two set differently)
Or that the rollback handler not do anything when the tableName and temporaryTableName are identical? (that would seem weird)
Won't any derived `beginInsertTable` now equally have to be aware of that or have the table removed?

Happy to do that as long as we can fix the problem.

Reply:
How about this?